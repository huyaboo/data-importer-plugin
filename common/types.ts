import { CSV_FILE_TYPE, JSON_FILE_TYPE, NDJSON_FILE_TYPE } from './constants';

export type SupportedFileTypes =
  | typeof CSV_FILE_TYPE
  | typeof JSON_FILE_TYPE
  | typeof NDJSON_FILE_TYPE;
