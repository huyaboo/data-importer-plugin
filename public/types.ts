import { DataSourceManagementPluginSetup } from 'src/plugins/data_source_management/public';
import { DataSourcePluginSetup } from 'src/plugins/data_source/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

export interface DataImporterPluginSetupDeps {
  dataSourceManagement?: DataSourceManagementPluginSetup;
  dataSource?: DataSourcePluginSetup;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataImporterPluginSetup {}
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataImporterPluginStart {}

export interface DataImporterPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
}

export interface ImportResponse {
  message: {
    total: number;
    message: string;
  };
  success: boolean;
}
