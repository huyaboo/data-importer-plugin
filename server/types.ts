import { OpenSearchClient } from 'opensearch-dashboards/server';
import { Readable } from 'stream';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataImporterPluginSetup {}
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataImporterPluginStart {}

export interface IngestOptions {
  client: OpenSearchClient;
  indexName: string;
  delimiter?: string;
  dataSourceId?: string;
}

export interface IngestResponse {
  total: number;
  message: string;
}

export interface ValidationOptions {
  delimiter?: string;
}

export interface IFileParser {
  validateText: (text: string, options: ValidationOptions) => Promise<boolean>;
  ingestFile: (file: Readable, options: IngestOptions) => Promise<IngestResponse>;
  ingestText: (text: string, options: IngestOptions) => Promise<IngestResponse>;
}
