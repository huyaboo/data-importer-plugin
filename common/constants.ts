export const CSV_FILE_TYPE = 'csv';
export const JSON_FILE_TYPE = 'json';
export const NDJSON_FILE_TYPE = 'ndjson';

export const SUPPORTED_FILE_TYPES = [CSV_FILE_TYPE, JSON_FILE_TYPE, NDJSON_FILE_TYPE];
