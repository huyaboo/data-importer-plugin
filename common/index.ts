export const PLUGIN_ID = 'staticDataIngestion';
export const PLUGIN_NAME = 'staticDataIngestion';
