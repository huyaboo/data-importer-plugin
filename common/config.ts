import { schema, TypeOf } from '@osd/config-schema';
import { CSV_FILE_TYPE, JSON_FILE_TYPE, NDJSON_FILE_TYPE } from './constants';

export const configSchema = schema.object({
  enabled: schema.boolean({ defaultValue: false }),
  enabledFileTypes: schema.arrayOf(schema.string(), {
    defaultValue: [CSV_FILE_TYPE, JSON_FILE_TYPE, NDJSON_FILE_TYPE],
  }),
  maxFileSizeBytes: schema.number({
    defaultValue: 100000000,
    min: 1,
  }),
  maxTextCount: schema.number({
    defaultValue: 10000,
    min: 1,
  }),
});

export type ConfigSchema = TypeOf<typeof configSchema>;
export type PublicConfigSchema = Omit<ConfigSchema, 'enabled'>;
