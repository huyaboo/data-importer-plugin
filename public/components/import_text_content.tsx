import React, { useState } from 'react';

import {
  EuiCodeEditor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPageContentBody,
  EuiSelect,
  EuiText,
} from '@elastic/eui';
import { SupportedFileTypes } from '../../common/types';
import { JSON_FILE_TYPE } from '../../common/constants';

export interface ImportTextContentBodyProps {
  onTextChange: (text: string) => void;
  enabledFileTypes: SupportedFileTypes[];
  onFileTypeChange: (fileType: SupportedFileTypes) => void;
  characterLimit: number;
  initialFileType: SupportedFileTypes;
}

export const ImportTextContentBody = ({
  onTextChange,
  enabledFileTypes,
  onFileTypeChange,
  characterLimit,
  initialFileType,
}: ImportTextContentBodyProps) => {
  const [codeEditorText, setCodeEditorText] = useState<string>('');
  const [fileType, setFileType] = useState<SupportedFileTypes>(initialFileType);
  const [numCharacters, setNumCharacters] = useState<number>(0);
  const options = enabledFileTypes.map((type) => {
    return {
      value: type,
      text: type,
    };
  });

  const onTextUpdate = (text: string) => {
    setCodeEditorText(text);
    onTextChange(text);
    setNumCharacters(text.length);
  };

  const onOptionSelect = (e) => {
    setFileType(e.target.value === JSON_FILE_TYPE ? e.target.value : undefined);
    onFileTypeChange(e.target.value as SupportedFileTypes);
  };

  return (
    <EuiPageContentBody>
      <EuiFlexGroup direction="column">
        <EuiFlexItem>
          <EuiFlexGroup justifyContent="flexStart" gutterSize="s">
            <EuiFlexItem grow={false}>
              <EuiText>File Format: </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiSelect
                options={options}
                fullWidth={false}
                onChange={onOptionSelect}
                value={initialFileType}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiCodeEditor
            onChange={onTextUpdate}
            width={'full'}
            value={codeEditorText}
            mode={fileType}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiText color={numCharacters > characterLimit ? 'danger' : 'default'}>
            {numCharacters}/{characterLimit} characters
          </EuiText>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiPageContentBody>
  );
};
