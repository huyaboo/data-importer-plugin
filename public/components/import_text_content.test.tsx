import React from 'react';
import { render } from '@testing-library/react';
import { ImportTextContentBody } from './import_text_content';
import { SupportedFileTypes } from '../../common/types';
import { SUPPORTED_FILE_TYPES_LIST } from '../../common';

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'mocked-id'),
}));

describe('ImportTextContent', () => {
  it('should render', () => {
    const { container } = render(
      <ImportTextContentBody
        onTextChange={jest.fn()}
        enabledFileTypes={SUPPORTED_FILE_TYPES_LIST as SupportedFileTypes[]}
        onFileTypeChange={jest.fn()}
        characterLimit={10000}
        initialFileType={SUPPORTED_FILE_TYPES_LIST[0] as SupportedFileTypes}
      />
    );

    // These id fields cannot be manually set
    container.querySelector('div.ace_editor')?.removeAttribute('id');
    container.querySelector('button.euiCodeEditorKeyboardHint')?.removeAttribute('id');

    expect(container).toMatchSnapshot();
  });
});
