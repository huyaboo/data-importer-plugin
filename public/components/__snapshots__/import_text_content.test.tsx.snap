// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportTextContent should render 1`] = `
<div>
  <div
    class="euiPageContentBody"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--medium"
            >
              File Format: 
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <select
                  class="euiSelect"
                >
                  <option
                    value="csv"
                  >
                    csv
                  </option>
                  <option
                    value="json"
                  >
                    json
                  </option>
                  <option
                    value="ndjson"
                  >
                    ndjson
                  </option>
                </select>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem"
      >
        <div
          class="euiCodeEditorWrapper"
          data-test-subj="codeEditorContainer"
        >
          <button
            class="euiCodeEditorKeyboardHint"
            data-test-subj="codeEditorHint"
          >
            <p
              class="euiText"
            >
              Press Enter to start editing.
            </p>
            <p
              class="euiText"
            >
              When you're done, press Escape to stop editing.
            </p>
          </button>
          <div
            class=" ace_editor ace-tm"
            style="height: 500px;"
          >
            <textarea
              autocapitalize="off"
              autocorrect="off"
              class="ace_text-input"
              id="mocked-id"
              spellcheck="false"
              style="opacity: 0;"
              tabindex="-1"
              wrap="off"
            />
            <div
              aria-hidden="true"
              class="ace_gutter"
            >
              <div
                class="ace_layer ace_gutter-layer ace_folding-enabled"
              />
              <div
                class="ace_gutter-active-line"
              />
            </div>
            <div
              class="ace_scroller"
            >
              <div
                class="ace_content"
              >
                <div
                  class="ace_layer ace_print-margin-layer"
                >
                  <div
                    class="ace_print-margin"
                    style="left: 4px; visibility: visible;"
                  />
                </div>
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_text-layer"
                  style="padding: 0px 4px;"
                />
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_cursor-layer ace_hidden-cursors"
                >
                  <div
                    class="ace_cursor"
                  />
                </div>
              </div>
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-v"
              style="display: none; width: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="width: 20px;"
              />
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-h"
              style="display: none; height: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="height: 20px;"
              />
            </div>
            <div
              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
            >
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              />
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              >
                XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem"
      >
        <div
          class="euiText euiText--medium"
        >
          <div
            class="euiTextColor euiTextColor--default"
          >
            0
            /
            10000
             characters
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
